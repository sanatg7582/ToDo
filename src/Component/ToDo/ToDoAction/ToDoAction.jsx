import React from "react";
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrash, faPen } from "@fortawesome/free-solid-svg-icons";
import Button from "../../Shared/Button";
/**
 * This function return Edit and Delete button
 * @param {number} index
 * @param {func} handleOnClickRemove
 * @param {func} handleOnClickUpdate
 * @returns Button Element
 */
const ToDoAction = props => {
  const { index, handleOnClickRemove, handleOnClickUpdate } = props;
  return (
    <>
      <Button
        buttonType="button"
        btnClass="warning"
        handleOnClick={handleOnClickUpdate}
        index={index}
      >
        <FontAwesomeIcon icon={faPen} />
      </Button>
      &nbsp;
      <Button
        buttonType="button"
        btnClass="danger"
        handleOnClick={handleOnClickRemove}
        index={index}
      >
        <FontAwesomeIcon icon={faTrash} />
      </Button>
    </>
  );
};
// default props value
ToDoAction.defaultProps = {
  index: 0,
  handleOnClickRemove: () => {},
  handleOnClickUpdate: () => {}
};
//prop type validation
ToDoAction.propTypes = {
  index: PropTypes.number,
  handleOnClickRemove: PropTypes.func.isRequired,
  handleOnClickUpdate: PropTypes.func.isRequired
};
export default ToDoAction;
