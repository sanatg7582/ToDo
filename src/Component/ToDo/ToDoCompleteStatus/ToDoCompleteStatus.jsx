import React from "react";
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faCheckCircle,
  faTimesCircle
} from "@fortawesome/free-solid-svg-icons";
import Button from "../../Shared/Button";

/**
 * This function return Is complete Button
 * @param {bool} isCompleteStatus,
 * @param {func} handleOnClick,
 * @param {number} index
 * @returns Element
 */
const ToDoCompleteStatus = props => {
  const { isCompleteStatus, handleOnClick, index } = props;
  return (
    <>
      {isCompleteStatus ? (
        <Button
          buttonType="button"
          btnClass="success"
          handleOnClick={handleOnClick}
          index={index}
          status={true}
        >
          <FontAwesomeIcon icon={faCheckCircle} />
        </Button>
      ) : (
        <Button
          buttonType="button"
          btnClass="danger"
          handleOnClick={handleOnClick}
          index={index}
          status={false}
        >
          <FontAwesomeIcon icon={faTimesCircle} />
        </Button>
      )}
    </>
  );
};
// default props value
ToDoCompleteStatus.defaultProps = {
  isCompleteStatus: false,
  handleOnClick: () => {},
  index: 0
};
// prop type validation
ToDoCompleteStatus.propTypes = {
  isCompleteStatus: PropTypes.bool.isRequired,
  handleOnClick: PropTypes.func.isRequired,
  index: PropTypes.number
};
export default ToDoCompleteStatus;
