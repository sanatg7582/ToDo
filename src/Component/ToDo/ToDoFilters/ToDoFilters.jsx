import React from "react";
import PropTypes from "prop-types";
import styles from "./ToDoFilters.module.scss";
import Button from "../../Shared/Button";
import Input from "../../Shared/Input";
/**
 * This function for return filter button
 * @param {func} handleOnClick
 * @param {number} filterActive
 * @param {func} setSearchtermfun
 * @param {func} searchterm
 * @returns Element
 */
const ToDoFilters = props => {
  const { handleOnClick, filterActive, setSearchtermfun, searchterm } = props;
  return (
    <div className={styles.filterContainer}>
      <div className={styles.buttonSection} >
        <Button
          buttonType="button"
          buttonName="All"
          btnClass={filterActive === 0 ? "success" : "primary"}
          handleOnClick={() => handleOnClick()}
          status={filterActive === 0 ? true : false}
          index={0}
        />
        &nbsp;&nbsp;&nbsp;&nbsp;
      <Button
          buttonType="button"
          buttonName="Active"
          btnClass={filterActive === 1 ? "success" : "primary"}
          handleOnClick={() => handleOnClick(1)}
          status={filterActive === 1 ? true : false}
          index={1}
        />
        &nbsp;&nbsp;&nbsp;&nbsp;
      <Button
          buttonType="button"
          buttonName="Is Completed"
          btnClass={filterActive === 2 ? "success" : "primary"}
          handleOnClick={() => handleOnClick(2)}
          status={filterActive === 2 ? true : false}
          index={2}
        />
      </div>
      <div className={styles.searchSection}>
        <Input InputType="text" InputName="searchterm" InputValue={searchterm} placeholder="search" handleChange={(e) => { setSearchtermfun(e.target.value) }} />
      </div>
    </div>
  );
};
// default props value
ToDoFilters.defaultProps = {
  handleOnClick: () => { },
  filterActive: 0
};
// prop type validation
ToDoFilters.propTypes = {
  handleOnClick: PropTypes.func.isRequired,
  filterActive: PropTypes.number
};
export default ToDoFilters;
