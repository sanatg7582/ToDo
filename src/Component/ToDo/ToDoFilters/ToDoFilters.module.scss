.filterContainer {
  padding: 10px 40px;
  display: flex;
  justify-content: space-between;
  .buttonSection{
    width: 75%;
  }
  .searchSection{
    width: 50%;
    margin-right: 68px;
  }
  input{
    width: 100%;
    margin-top: 10px;
  }
}
