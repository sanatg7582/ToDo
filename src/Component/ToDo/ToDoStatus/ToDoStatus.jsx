import React from "react";
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck, faTimes } from "@fortawesome/free-solid-svg-icons";
import Button from "../../Shared/Button";
/**
 * This function ToDo status button
 * @param {bool} todoStatus
 * @param {func} handleOnClick
 * @param {number} index
 * @return element
 */
const ToDoStatus = props => {
  const { todoStatus, handleOnClick, index } = props;
  return (
    <>
      {todoStatus ? (
        <Button
          buttonType="button"
          btnClass="success"
          handleOnClick={handleOnClick}
          index={index}
          status={true}
        >
          <FontAwesomeIcon icon={faCheck} />
        </Button>
      ) : (
        <Button
          buttonType="button"
          btnClass="danger"
          handleOnClick={handleOnClick}
          index={index}
          status={false}
        >
          <FontAwesomeIcon icon={faTimes} />
        </Button>
      )}
    </>
  );
};
// default props value
ToDoStatus.defaultProps = {
  todoStatus: true,
  handleOnClick: () => {},
  index: 0
};
// prop type validation
ToDoStatus.propTypes = {
  todoStatus: PropTypes.bool.isRequired,
  handleOnClick: PropTypes.func.isRequired,
  index: PropTypes.number
};
export default ToDoStatus;
