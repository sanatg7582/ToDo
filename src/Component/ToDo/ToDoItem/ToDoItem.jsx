import React from "react";
import PropTypes from "prop-types";
import Text from "../../Shared/Text";
import ToDOStatus from "../ToDoStatus";
import ToDoCompleteStatus from "../ToDoCompleteStatus";
import ToDoAction from "../ToDoAction";
/**
 * This function for todo Item row in table
 * @param {number} index
 * @param {string} todoHeading
 * @param {bool} todoStatus
 * @param {bool} todoIsCompleteStatus
 * @param {func} updateToDoStatus
 * @param {func} completeToDoStatus
 * @param {func} removeToDo
 * @param {func} updateToDo
 * @returns tr element
 */
const ToDoItem = props => {
  const {
    index,
    todoHeading,
    todoStatus,
    todoIsCompleteStatus,
    updateToDoStatus,
    completeToDoStatus,
    removeToDo,
    updateToDo
  } = props;
  return (
    <tr>
      <td>
        <Text strong={true} heading={todoHeading} />
      </td>
      <td>
        <ToDOStatus
          todoStatus={todoStatus}
          handleOnClick={updateToDoStatus}
          index={index}
        />
      </td>
      <td>
        <ToDoCompleteStatus
          isCompleteStatus={todoIsCompleteStatus}
          handleOnClick={completeToDoStatus}
          index={index}
        />
      </td>
      <td>
        <ToDoAction
          index={index}
          handleOnClickRemove={removeToDo}
          handleOnClickUpdate={updateToDo}
        />
      </td>
    </tr>
  );
};
// default props value
ToDoItem.defaultProps = {
  index: 0,
  todoHeading: "",
  todoStatus: true,
  todoIsCompleteStatus: false,
  updateToDoStatus: () => {},
  completeToDoStatus: () => {},
  removeToDo: () => {},
  updateToDo: () => {}
};
//props type validation
ToDoItem.propTypes = {
  index: PropTypes.number,
  todoHeading: PropTypes.string.isRequired,
  todoStatus: PropTypes.bool.isRequired,
  todoIsCompleteStatus: PropTypes.bool.isRequired,
  updateToDoStatus: PropTypes.func.isRequired,
  completeToDoStatus: PropTypes.func.isRequired,
  removeToDo: PropTypes.func.isRequired,
  updateToDo: PropTypes.func.isRequired
};
export default ToDoItem;
