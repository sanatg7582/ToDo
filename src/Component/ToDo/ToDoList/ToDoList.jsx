import React, { useState } from "react";
import PropTypes from "prop-types";
import ToDoItem from "../ToDoItem";
import ToDoFilters from "../ToDoFilters";

/**
 * This component for render ToDoList
 * @param {array} todos // array of object
 * @param {func} updateToDos // func
 * @param {func} handleOnClick // func
 */

const ToDoList = (props) => {
  const [filterActive, setFilterActive] = useState(0);
  const [searchterm, setSearchterm] = useState();
  const { todos, updateToDos, handleOnClick } = props;

  // remove toDo
  const removeToDo = title => {
    updateToDos(
      todos.filter(itm => {
        return itm.todo !== title;
      })
    );
  };

  // Update toDo Status
  const updateToDoStatus = index => {
    updateToDos(
      todos.map((item, key) =>
        key === index ? { ...item, status: !item.status } : item
      )
    );
  };

  // Update complete ToDo status
  const completeToDoStatus = index => {
    updateToDos(
      todos.map((itm, key) =>
        key === index ? { ...itm, isCompleted: !itm.isCompleted } : itm
      )
    );
  };

  // update ToDo
  const updateToDo = (e, index) => {
    handleOnClick(e, index);
  };


  let newtoDos = [...todos];
  if (filterActive === 1) {
    newtoDos = todos.filter(itm => {
      return itm.status === true
    });
  }
  if (filterActive === 2) {
    newtoDos = todos.filter(itm => {
      return itm.isCompleted === true
    });
  }
  if (searchterm && searchterm.length > 0) {
    newtoDos = newtoDos.filter(itm => itm.todo.indexOf(searchterm) >= 0);
  }

  return (
    <>
      <ToDoFilters
        handleOnClick={(status) => { setFilterActive(status) }}
        filterActive={filterActive}
        searchterm={searchterm}
        setSearchtermfun={(key) => { setSearchterm(key) }}
      />

      <table>
        <thead>
          <tr>
            <th>ToDo</th>
            <th>Status</th>
            <th>IsCompleted</th>
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          {newtoDos.map((todo, index) => {
            return (
              <ToDoItem
                index={index}
                todoHeading={todo.todo}
                todoStatus={todo.status}
                todoIsCompleteStatus={todo.isCompleted}
                updateToDoStatus={() => updateToDoStatus(index)}
                completeToDoStatus={() => completeToDoStatus(index)}
                removeToDo={() => removeToDo(todo.todo)}
                updateToDo={updateToDo}
                key={index}
              />
            );

          })}
        </tbody>
      </table>
    </>
  );

}
// default props 
ToDoList.defaultProps = {
  todos: [],
  updateToDos: () => { },
  handleOnClick: () => { }
}
// PropTypes validation
ToDoList.propTypes = {
  todos: PropTypes.array,
  updateToDos: PropTypes.func,
  handleOnClick: PropTypes.func
};
export default ToDoList;
