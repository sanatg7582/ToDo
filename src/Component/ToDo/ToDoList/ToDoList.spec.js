import React from "react";
import { mount } from "enzyme";
import ToDoList from "./ToDoList";
import ToDoStatus from "../ToDoStatus";
import ToDoCompleteStatus from "../ToDoCompleteStatus";
import ToDoAction from "../ToDoAction";
import ToDoFilters from "../ToDoFilters";
import ToDoItem from "../ToDoItem";
describe("Load <ToDoList />", () => {
  let wrapper = "";
  beforeAll(() => {
    const props = {
      todos: [{ todo: "test", status: true, isCompleted: false }]
    };
    wrapper = mount(<ToDoList {...props} />);
  });
  it("renders without crashing", () => {
    expect(wrapper).toBeTruthy();
  });

  it("Checking length", () => {
    expect(wrapper).toHaveLength(1);
  });

  it("checking All toDo Filter button ", () => {
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(0)
        .hasClass("btn success")
    ).toEqual(true);
    wrapper
      .find(ToDoFilters)
      .find("button")
      .at(0)
      .simulate("click");
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(0)
        .hasClass("btn success")
    ).toEqual(true);
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(1)
        .hasClass("btn primary")
    ).toEqual(true);
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(2)
        .hasClass("btn primary")
    ).toEqual(true);
  });

  it("checking Active toDo Filter button ", () => {
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(1)
        .hasClass("btn primary")
    ).toEqual(true);
    wrapper
      .find(ToDoFilters)
      .find("button")
      .at(1)
      .simulate("click");
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(0)
        .hasClass("btn primary")
    ).toEqual(true);
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(1)
        .hasClass("btn success")
    ).toEqual(true);
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(2)
        .hasClass("btn primary")
    ).toEqual(true);
  });

  it("checking IsCompleted toDo Filter button ", () => {
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(2)
        .hasClass("btn primary")
    ).toEqual(true);
    wrapper
      .find(ToDoFilters)
      .find("button")
      .at(2)
      .simulate("click");
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(0)
        .hasClass("btn primary")
    ).toEqual(true);
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(1)
        .hasClass("btn primary")
    ).toEqual(true);
    expect(
      wrapper
        .find(ToDoFilters)
        .find("button")
        .at(2)
        .hasClass("btn success")
    ).toEqual(true);
  });

  it("checking table <th> mock data", () => {
    expect(
      wrapper
        .find("table")
        .find("thead")
        .find("tr")
        .find("th")
        .first()
        .text()
    ).toEqual("ToDo");
    expect(
      wrapper
        .find("table")
        .find("thead")
        .find("tr")
        .find("th")
        .at(1)
        .text()
    ).toEqual("Status");
    expect(
      wrapper
        .find("table")
        .find("thead")
        .find("tr")
        .find("th")
        .at(2)
        .text()
    ).toEqual("IsCompleted");
    expect(
      wrapper
        .find("table")
        .find("thead")
        .find("tr")
        .find("th")
        .at(3)
        .text()
    ).toEqual("Action");
  });

  // it("checking ToDoStatus button onclick change class", () => {
  //   const props = {
  //     todos: [{ todo: "test", status: true, isCompleted: false }]
  //   };
  //   wrapper = mount(<ToDoList {...props} />);
  //   expect(
  //     wrapper
  //       .find(ToDoStatus)
  //       .find("button")
  //       .hasClass("btn success")
  //   ).toEqual(true);
  //   wrapper
  //     .find(ToDoStatus)
  //     .find("button")
  //     .simulate("click");
  //   expect(
  //     wrapper
  //       .find(ToDoStatus)
  //       .find("button")
  //       .hasClass("btn danger")
  //   ).toEqual(true);
  //   expect(wrapper.find(ToDoItem)).toHaveLength(1);
  // });

  // it("checking ToDoCompleteStatus button onclick change class", () => {
  //   const props = {
  //     todos: [{ todo: "test", status: true, isCompleted: false }]
  //   };
  //   wrapper = mount(<ToDoList {...props} />);
  //   expect(
  //     wrapper
  //       .find(ToDoCompleteStatus)
  //       .find("button")
  //       .hasClass("btn danger")
  //   ).toEqual(true);
  //   wrapper
  //     .find(ToDoCompleteStatus)
  //     .find("button")
  //     .simulate("click");
  //   expect(
  //     wrapper
  //       .find(ToDoCompleteStatus)
  //       .find("button")
  //       .hasClass("btn success")
  //   ).toEqual(true);
  //   expect(wrapper.find(ToDoItem)).toHaveLength(1);
  // });

  it("checking ToDoAction button class", () => {
    const props = {
      todos: [{ todo: "test", status: true, isCompleted: false }]
    };
    wrapper = mount(<ToDoList {...props} />);
    expect(
      wrapper
        .find(ToDoAction)
        .find("button")
        .first()
        .hasClass("btn warning")
    ).toEqual(true);
    expect(
      wrapper
        .find(ToDoAction)
        .find("button")
        .last()
        .hasClass("btn danger")
    ).toEqual(true);
  });

  // it("checking remove todo click", () => {
  //   const props = {
  //     todos: [{ todo: "test", status: true, isCompleted: false }]
  //   };
  //   wrapper = mount(<ToDoList {...props} />);
  //   wrapper
  //     .find(ToDoAction)
  //     .find("button")
  //     .last()
  //     .simulate("click");
  //   expect(wrapper.find(ToDoItem)).toHaveLength(1);
  // });

  it("SnapShot test", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
