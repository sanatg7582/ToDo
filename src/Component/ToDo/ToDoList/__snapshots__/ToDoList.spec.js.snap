// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load <ToDoList /> SnapShot test 1`] = `ReactWrapper {}`;
